rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,182 @@
+
+- For stable releases:
+  - `pname` reflects the software name;
+  - `version` reflects the released version.
+- For unstable releases:
+  - `pname` reflects the software name;
+  - `version` is a string following the format `unstable-YYYY-MM-DD`, where
+    `YYYY-MM-DD` denotes the date when the code was generated.
+
+This is a simple and easy-to-understand format.
+
+However, it does not map very well with the Nix function
+`builtins.parseDrvName`:
+
+```example
+# expected: { name = "mpv"; version = "0.35.15"; }
+nix-repl> builtins.parseDrvName "mpv-0.35.15"
+builtins.parseDrvName "mpv-0.35.15"
+{ name = "mpv"; version = "0.35.15"; }
+
+# expected: { name = "mpv"; version = "unstable-2021-05-03"; }
+nix-repl> builtins.parseDrvName "mpv-unstable-2021-05-03"
+builtins.parseDrvName "mpv-unstable-2021-05-03"
+{ name = "mpv-unstable"; version = "2021-05-03"; }
+```
+
+It happens because the `version` attribute returned by `builtins.parseDrvName`
+starts with a digit. This is not a bug, but a deliberate design decision of Nix
+specification; therefore we should strive to follow it, neither circumventing
+nor ignoring it.
+
+That being said, `version` attribute should start with a digit.
+
+Furthermore, `version` attribute should follow the semantics for upgrading as
+stated in `nix-env` manpage.
+
+This document describes a format suitable to fix this issue while striving for
+simplicity.
+
+# Detailed design
+[design]: #detailed-design
+
+## Terminology
+
+_Disclaimer_: whereas some of the terms enumerated below are borrowed from
+[pkgsrc guide](https://www.netbsd.org/docs/pkgsrc/) plus a bit of terminology
+employed by git, this document aims to be general, not being overly attached to
+them.
+
+- Program denotes the piece of software to be fetched and processed via Nixpkgs.
+  - This term makes no distinction about the immediate or intended uses of the
+    program; it can range from non-interactive programs to full GUI
+    applications, even including filesets intended as input to other programs
+    (such as firmwares and fonts).
+- Team denotes the maintainers of the program.
+  - This term makes no distinction among the various models of organization of a
+    team; it ranges from a solitary programmer to a business company - and
+    everything inbetween.
+- Source denotes the origin of the program.
+  - This term makes no distinction among precompiled, binary-only or high-level
+    code distributions.
+- Snapshot denotes the source of the program taken at a point of the time.
+  - Labeled snapshot denotes any snapshot explicitly labeled by its team.
+  - Unlabeled snapshot denotes any snapshot not explicitly labeled by its team.
+- Release denotes any distributed snapshot, as defined by its team.
+- Branch denotes a logical sequence of snapshots, as identified by the program's
+  team;
+  - Usually these branches are denoted by names such like `stable`, `master`,
+    `unstable`, `trunk`, `experimental`, `staging` etc.;
```

Comment:
or by major version?

Reply:
A major version is more akin to a labeled snapshot, not a logical sequence of them. Indeed, major versions don't necessarily follow a continuous sequence. An example is Zig: between 0.8.0 and 0.8.1 there are many small, discrete snapshots.

Reply:
These only differ in patchlevel anyway. 

I mean that PostgreSQL has branches 13.x and 14.x and some more.

Linux kernel, annoyingly, has stable branches differing just in minor version (although major.minor is just formatting for Linux kernel versions). So maybe a branch can be called by a version prefix instead of a name?